rust/kernel/miscdev.rs

```diff
@@ -7,91 +7,97 @@
+unsafe impl<T: Sync> Sync for Registration<T> {}
+
+impl<T: Sync> Registration<T> {
     /// Creates a new [`Registration`] but does not register it yet.
     ///
     /// It is allowed to move.
-    pub fn new() -> Self {
+    pub fn new(context: T) -> Self {
         Self {
-            mdev: None,
+            registered: false,
+            mdev: bindings::miscdevice::default(),
             _pin: PhantomPinned,
+            context,
         }
     }
 
     /// Registers a miscellaneous device.
     ///
     /// Returns a pinned heap-allocated representation of the registration.
-    pub fn new_pinned<T: FileOperations>(
+    pub fn new_pinned<F: FileOpener<T>>(
         name: CStr<'static>,
         minor: Option<i32>,
+        state: T,
     ) -> KernelResult<Pin<Box<Self>>> {
-        let mut r = Pin::from(Box::try_new(Self::new())?);
-        r.as_mut().register::<T>(name, minor)?;
+        let mut r = Pin::from(Box::try_new(Self::new(state))?);
+        r.as_mut().register::<F>(name, minor)?;
         Ok(r)
     }
 
     /// Registers a miscellaneous device with the rest of the kernel.
     ///
-    /// It must be pinned because the memory block that represents the
-    /// registration is self-referential. If a minor is not given, the kernel
-    /// allocates a new one if possible.
-    pub fn register<T: FileOperations>(
+    /// It must be pinned because the memory block that represents the registration is
+    /// self-referential. If a minor is not given, the kernel allocates a new one if possible.
+    pub fn register<F: FileOpener<T>>(
         self: Pin<&mut Self>,
         name: CStr<'static>,
         minor: Option<i32>,
     ) -> KernelResult {
         // SAFETY: We must ensure that we never move out of `this`.
         let this = unsafe { self.get_unchecked_mut() };
-        if this.mdev.is_some() {
+        if this.registered {
             // Already registered.
             return Err(Error::EINVAL);
         }
 
-        this.mdev = Some(bindings::miscdevice::default());
-        let dev = this.mdev.as_mut().unwrap();
-        dev.fops = &FileOperationsVtable::<T>::VTABLE;
-        dev.name = name.as_ptr() as *const c_types::c_char;
-        dev.minor = minor.unwrap_or(bindings::MISC_DYNAMIC_MINOR as i32);
-        let ret = unsafe { bindings::misc_register(dev) };
+        this.mdev.fops = &FileOperationsVtable::<Self, F>::VTABLE;
+        this.mdev.name = name.as_ptr() as *const c_types::c_char;
+        this.mdev.minor = minor.unwrap_or(bindings::MISC_DYNAMIC_MINOR as i32);
+
+        let ret = unsafe { bindings::misc_register(&mut this.mdev) };
         if ret < 0 {
-            this.mdev = None;
             return Err(Error::from_kernel_errno(ret));
         }
+        this.registered = true;
         Ok(())
     }
 }
 
-impl Default for Registration {
-    fn default() -> Self {
-        Self::new()
+impl<T: Sync> FileOpenAdapter for Registration<T> {
```

Comment:
I'm still thinking my way through this, but does this impl potentially allow you to use a `Registration<T>` as the adapter type with a _non_ miscdev file, leading to a type confusion issue in `convert`?

Reply:
Good catch! It could. But only within the `kernel` crate because `FileOperationsVtable` is not accessible out of it.

Still not good. Let me fix this.

Reply:
Fixed with the latest commit.